features/steps/steps.py

```diff
@@ -200,6 +234,65 @@ def step_impl(context, constraint, num, entity):
 
     handle_errors(context, errors)
 
+@then('Each {entity} must nest {constraint} {num:d} instance(s) of {other_entity}')
+def step_impl(context, entity, num, constraint, other_entity):
+    stmt_to_op = {'exactly': operator.eq, "at most": operator.le}
+    assert constraint in stmt_to_op
+    op = stmt_to_op[constraint]
+
+    errors = []
+
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            nested_entities = [entity for rel in inst.IsNestedBy for entity in rel.RelatedObjects]
+            if not op(len([1 for i in nested_entities if i.is_a() == other_entity]), num):
+                errors.append(instance_structure_error(inst, [i for i in nested_entities if i.is_a() == other_entity], 'nesting'))
+                # errors.append(instance_count_error([i for i in nested_entities if i.is_a() == other_entity]))
+
+
+    handle_errors(context, errors)
+
+@then('Each {entity} must be nested only by {num:d} {other_entity}')
+def step_impl(context, entity, other_entity, num):
+    errors = []
+
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            relating = [i.RelatingObject for i in inst.Nests]
+            if not all([len(relating) <= num, other_entity == relating[0].is_a()]):
```

Comment:
This is a bit peculiar.

`inst.Nests` is length 0 or 1 in the schema already. So I would propose:

- the Gherkin rule should be rewritten without the `num:d` (it becomes a duplicate then of `.. nests a list of only ..`?
- if we have a list of length 0..1 I'd wouldn't use the `all()` predicate, especially since you already only do `relating[0]`.

Reply:
Implemented in https://github.com/buildingSMART/ifc-gherkin-rules/commit/0c8fdcad7ba2cb8627485193daa2643b93e770a3
The rule is rewritten without the num:d and is indeed a duplicate. The implementation is also more general, so it will check for both more than one relationship as well as more relating objects within a relationship. If the gherkin sentence is specific however, as 'nests a list ..', the code will check for just a list.